Packs/Hackuity/Integrations/Hackuity/Hackuity.py

```diff
@@ -0,0 +1,694 @@
+                message=f"Authentication error: {error_description}",
+                exception=e.exception,
+                res=e.res,
+            ) from e
+
+        set_integration_context(
+            {
+                # clear other keys to force refresh
+                "access_token": response["access_token"],
+                "access_token_expiration": int(time.time()) + response["expires_in"],
+            }
+        )
+        return response["access_token"]
+
+    def get_user_id(self) -> str:
+        """Get the id of the user"""
+        integration_context = get_integration_context()
+        user_id = integration_context.get("user_id", {})
+
+        if not user_id:
+            user_id = self.fetch_user()["userId"]
+            integration_context["user_id"] = user_id
+            set_integration_context(integration_context)
+
+        return user_id
+
+    def get_vulnerability_name(self, vuln_id: str) -> Optional[str]:
+        """Get the name of a vulnerability from its id"""
+        # read from cache
+        integration_context = get_integration_context()
+        vulnerability_names = integration_context.get("vulnerability_names", {})
+        if vuln_id in vulnerability_names:
+            return vulnerability_names[vuln_id]
+
+        # fetch vulnerability
+        try:
+            result = self.http_request(
+                url_suffix=f"/namespaces/N000000101010/vulnerabilityTypes/{vuln_id}"
+            )
+            name = result["i18n"]["en"]["name"]
+        except DemistoException:
+            return None  # failed to get the name (don't save it)
+
+        # save into cache
+        vulnerability_names[vuln_id] = name
+        integration_context["vulnerability_names"] = vulnerability_names
+        set_integration_context(integration_context)
+        return name
+
+    def compute_asset_id(self, asset_name: str, asset_type: Optional[str]) -> str:
+        request: Dict[str, Any] = {"assetAbsoluteNames": [asset_name]}
+        if asset_type:
+            request["assetTypes"] = [asset_type.upper()]
+        response = self.http_request(
+            method="POST",
+            url_suffix="/search/assets",
+            json_data=request,
+        )
+        for asset in response.get("searchAssets", []):
+            if asset["assetAbsoluteName"] == asset_name:
+                return asset["assetId"]
+        raise ValueError("Asset not found")
+
+    def fetch_dashboard_widgets(self) -> List[Dict[str, Any]]:
+        response = self.http_request(
+            method="GET",
+            url_suffix=(
+                f"/namespaces/{self.namespace}/dashboards/{self.get_user_id()}"
+            ),
+        )
+        return [
+            json_extract(
+                widget,
+                {
+                    "id": ["id"],
+                    "params": ["params", None],
+                    "type": ["type"],
+                },
+            )
+            for widget in response.get("widgets", [])
```

Comment:
if you're trying to get the json object of each widget, why dont you just return it as it is from the api, or use an already existing methods.

Reply:
Because my `http_request` passes a `resp_type` of value `json` by default to `BaseClient._http_request`, the `response` object will already be json-decoded.


---

The `None` value in the `path` arg of `json_get_value` means “get the first value you find”.

The response returned by the server looks like this (some extra fields have been removed):
```json
{
  "kind": "hy#dashboard",
  "widgets": [
    {
      "id": "abc123",
      "params": {
        "overallScore": "redacted for clarity #1"
      },
      "type": "OVERALL_SCORE"
    },
    {
      "id": "def456",
      "params": {
        "scoreMeter": "redacted for clarity #2"
      },
      "type": "SCORE_METER"
    }
  ]
}
```

And we are only looking to return the following (note that we removed the nesting for the `params` keys):
```json
[
  {
    "id": "abc123",
    "params": "redacted for clarity #1",
    "type": "OVERALL_SCORE"
  },
  {
    "id": "def456",
    "params": "redacted for clarity #2",
    "type": "SCORE_METER"
  }
]
```